discovery/file/file.go

```diff
@@ -35,6 +37,71 @@ import (
 
 const fileSDFilepathLabel = model.MetaLabelPrefix + "filepath"
 
+// TimestampCollector is a Custom Collector for Timestamps of the files.
+type TimestampCollector struct {
+	index       int
+	filenames   map[string]struct{}
+	Description *prometheus.Desc
+	lock        sync.RWMutex
+
+	logger log.Logger
+}
+
+// Describe method sends the description to the channel.
+func (t *TimestampCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- t.Description
+}
+
+// SetFiles adds the filenames of the struct's map to the paths returned by listfiles().
+func (t *TimestampCollector) SetFiles(files []string) {
+	t.lock.Lock()
+	for _, file := range files {
+		t.filenames[file] = struct{}{}
+	}
+	t.lock.Unlock()
+}
+
+// DeleteFile deletes a key from the filenames map. Go's delete does nothing if the key does not exist.
+func (t *TimestampCollector) DeleteFile(file string) {
+	t.lock.Lock()
+	delete(t.filenames, file)
+	t.lock.Unlock()
+	level.Error(t.logger).Log("msg", "deleting file from timestamp collector", "file", file)
+}
+
+// Collect creates constant metrics for each file with last modified time of the file.
+func (t *TimestampCollector) Collect(ch chan<- prometheus.Metric) {
+	t.lock.RLock()
+	files := t.filenames
+	t.lock.RUnlock()
+	for k := range files {
+		info, err := os.Stat(k)
```

Comment:
We really should have the mtime at the last time we read it, not just what's currently on the filesystem. This would allow for detecting things being stuck.